lsp-diagnostics.el

```diff
@@ -0,0 +1,314 @@
+  (add-hook 'flymake-diagnostic-functions 'lsp-diagnostics--flymake-backend nil t)
+  (add-hook 'lsp-diagnostics-updated-hook 'lsp-diagnostics--flymake-after-diagnostics nil t))
+
+(defun lsp-diagnostics--flymake-after-diagnostics ()
+  "Handler for `lsp-diagnostics-updated-hook'."
+  (cond
+   ((and lsp-diagnostics--flymake-report-fn flymake-mode)
+    (lsp-diagnostics--flymake-update-diagnostics))
+   ((not flymake-mode)
+    (setq lsp-diagnostics--flymake-report-fn nil))))
+
+(defun lsp-diagnostics--flymake-backend (report-fn &rest _args)
+  "Flymake backend using REPORT-FN."
+  (let ((first-run (null lsp-diagnostics--flymake-report-fn)))
+    (setq lsp-diagnostics--flymake-report-fn report-fn)
+    (when first-run
+      (lsp-diagnostics--flymake-update-diagnostics))))
+
+(defun lsp-diagnostics--flymake-update-diagnostics ()
+  "Report new diagnostics to flymake."
+  (funcall lsp-diagnostics--flymake-report-fn
+           (-some->> (lsp-diagnostics t)
+             (gethash (lsp--fix-path-casing buffer-file-name))
+             (--map (-let* (((&Diagnostic :message :severity?
+                                          :range (range &as &Range
+                                                        :start (&Position :line start-line :character)
+                                                        :end (&Position :line end-line))) it)
+                            ((start . end) (lsp--range-to-region range)))
+                      (when (= start end)
+                        (if-let ((region (flymake-diag-region (current-buffer)
+                                                              (1+ start-line)
+                                                              character)))
+                            (setq start (car region)
+                                  end (cdr region))
+                          (lsp-save-restriction-and-excursion
+                            (goto-char (point-min))
+                            (setq start (point-at-bol (1+ start-line))
+                                  end (point-at-eol (1+ end-line))))))
+                      (flymake-make-diagnostic (current-buffer)
+                                               start
+                                               end
+                                               (cl-case severity?
+                                                 (1 :error)
+                                                 (2 :warning)
+                                                 (t :note))
+                                               message))))
+           ;; This :region keyword forces flymake to delete old diagnostics in
+           ;; case the buffer hasn't changed since the last call to the report
+           ;; function. See https://github.com/joaotavora/eglot/issues/159
+           :region (cons (point-min) (point-max))))
+
+
+
+;;;###autoload
+(defun lsp-diagnostics--enable ()
+  "Enable LSP checker support."
+  (unless (eq lsp-diagnostics-provider :none)
```

Comment:
nit: since the `:none` configuration is acceptable (user provide their own configuration), do you need to check this?
Shouldn't this check changed to something like `(when (member lsp-dianostics-provider '(:auto :none ....)))`?

Note that the mode is not turn on is also fine, but I can imagine that if user bind a key map for this mode and not have this turn on may cause problem.

Reply:
You mean only enable the mode if it was provided a known provider like `:auto`, `:none`, `:flycheck` or `:flymake`?

Reply:
Yes, `t` and `nil` too. Since those are configurations that don't trigger `lsp--warn`